tensorflow/core/kernels/dml_crop_and_resize_grad_boxes_op.cc

```diff
@@ -0,0 +1,207 @@
+  DML_INTERPOLATION_MODE GetInterpolationMode() const {
+    return attr_->interpolation_mode;
+  }
+
+ private:
+  TensorShape output_shape_;
+  std::shared_ptr<const Attributes> attr_;
+};
+
+class CropAndResizeGradBoxesShapeHelper : public ShapeHelper {
+ public:
+  std::vector<TensorShape> GetOutputShapes(
+      OpKernelContext* ctx,
+      const InitializationHelper* initialization_helper) const override {
+    auto init_helper = static_cast<const CropAndResizeGradBoxesInitHelper*>(
+        initialization_helper);
+
+    return {init_helper->GetOutputShape()};
+  }
+};
+
+class DmlCropAndResizeGradBoxesKernel : public DmlKernel {
+ public:
+  using InitHelper = CropAndResizeGradBoxesInitHelper;
+
+  explicit DmlCropAndResizeGradBoxesKernel(DmlKernelConstruction* ctx,
+                                           const InitHelper* init_helper) {
+    const TensorShape& image_shape = ctx->GetInputTensorShape(1);
+
+    DmlKernelParams params;
+    params.kernel_input_indices = {0, 1, 2, 3};
+
+    using namespace DmlTensorAxes;
+    auto layout = {N, H, W, C};
+
+    DmlKernelTensors tensors = GetTensorInfos(ctx, params);
+    tensors.inputs[0]->desc = CreateTensorDescFromInput(ctx, 0, layout);
+    tensors.inputs[1]->desc = CreateTensorDescFromInput(ctx, 1, layout);
+
+    // RoiAlignRoiGrad only supports uint32 for the batch indices
+    tensors.inputs[3]->desc.ForceUnsignedDataType();
+
+    auto inputs = GetDmlTensorDescs(tensors.inputs);
+    auto scope = dml::Graph(ctx->GetDmlDevice());
+    auto input_gradient = dml::InputTensor(scope, 0, inputs[0]);
+    auto image = dml::InputTensor(scope, 1, inputs[1]);
+    auto roi = dml::InputTensor(scope, 2, inputs[2]);
+    auto batch_indices = dml::InputTensor(scope, 3, inputs[3]);
+
+    if (ctx->GetInputDataType(0) != ctx->GetInputDataType(1)) {
+      auto dml_dtype = GetDmlDataTypeFromTfDataType(ctx->GetInputDataType(0));
+      image = dml::Cast(image, dml_dtype);
+    }
+
+    // TF's ROIs are in {y1, x1, y2, x2} order, but DML's are in {x1, y1, x2,
+    // y2} order
+    auto roi_sizes = roi.GetOutputDesc().sizes;
+    roi = dml::Reinterpret(roi, {1, roi_sizes[2], 2, 2}, {});
+
+    auto roi_seq_lengths =
+        dml::ScalarTensor<uint32_t>(scope, 2, {1, roi_sizes[2], 2, 1});
+    roi = dml::ReverseSubsequences(roi, roi_seq_lengths, 3);
+    roi = dml::Reinterpret(roi, roi_sizes, {});
+
+    const float spatial_scale_x = image_shape.dim_size(2) - 1;
+    const float spatial_scale_y = image_shape.dim_size(1) - 1;
+    constexpr uint32_t minimum_samples_per_output = 1;
+    constexpr uint32_t maximum_samples_per_output = 1;
+    constexpr float input_pixel_offset = 0;
+    constexpr float output_pixel_offset = 0;
+    constexpr bool align_region_to_corners = true;
+    const uint32_t batch_size = image_shape.dim_size(0);
+    const uint32_t image_height = image_shape.dim_size(1);
+    const uint32_t image_width = image_shape.dim_size(2);
+    constexpr bool computeOutputGradient = false;
```

Comment:
[nit] Personally I wish all the `snake_case` variables above were converted to `camelCase` :b, but so long as this code lives inside the TF repo, it's probably better for consistency to be like the other variables.